addons/website/models/website.py

```diff
@@ -1342,3 +1342,99 @@ def _get_cached_values(self):
+                AND f.model NOT LIKE 'ir.actions%'
+                AND f.model != 'mail.message'
+        """)
+        for model, name in cr.fetchall():
+            table = self.env[model]._table
+            if tools.table_exists(cr, table) and tools.column_exists(cr, table, name):
+                html_fields.append((table, name))
+        return html_fields
+
+    def _get_snippets_assets(self):
+        """Returns every snippet asset from the database in the form of a list :
+        [snippet_module, snippet_id, asset_version, asset_type, asset_id]
+        """
+        self.env.cr.execute(r"""
+            SELECT regexp_matches[1] AS snippet_module,
+                   regexp_matches[2] AS snippet_id,
+                   regexp_matches[3] AS asset_version,
+                   CASE WHEN regexp_matches[4]='scss' THEN 'css'
+                        ELSE regexp_matches[4]
+                   END AS asset_type,
+                   id AS asset_id
+            FROM
+                (SELECT REGEXP_MATCHES(PATH, '(\w*)\/.*\/snippets\/(\w*)\/(\d{3})\.(js|scss)'),
+                        id
+                FROM ir_asset
+                ORDER BY id ASC) AS regexp;
+        """)
+        return self.env.cr.fetchall()
+
+    def _is_snippet_used(self, snippet_module, snippet_id, asset_version, asset_type, html_fields):
+        snippet_occurences = []
+        # We check every snippet template definition
+        snippet_template = self.env.ref('%s.%s' % (snippet_module, snippet_id), raise_if_not_found=False)
+        if snippet_template:
+            snippet_template_html = snippet_template._render()
+            match = re.search('<([^>]*class="[^>]*)>', snippet_template_html)
+            snippet_occurences.append(match.group())
+
+        # As well as every snippet dropped in html fields
+        snippet_regex = '<([^>]*data-snippet="%s"[^>]*)>' % snippet_id
+        snippet_dropped = "UNION ".join("""SELECT REGEXP_MATCHES(%s, '%s') FROM %s """ % (column, snippet_regex, table) for table, column in html_fields)
+        self.env.cr.execute(snippet_dropped)
+        results = self.env.cr.fetchall()
+        for r in results:
+            snippet_occurences.append(r[0][0])
+
+        for snippet in snippet_occurences:
+            if asset_version == '000':
+                if 'data-v' + asset_type not in snippet:
+                    return True
+            else:
+                if 'data-v' + asset_type + '="' + asset_version + '"' in snippet:
+                    return True
+        return False
+
+    @api.autovacuum
+    def _disable_unused_snippets_assets(self):
+        snippets_assets = self._get_snippets_assets()
+        html_fields = self._get_html_fields()
+
+        # Some modules define overrides of assets. The parent asset
+        # and its overrides share the same snippet_id, asset_version and
+        # asset_type and must be activated/deactivated at the same time.
+        parent_assets = dict()
+        overrides_ids = dict()
```

Comment:
Cant we get rid of all this python code handling duplicate/override?
What I had in mind is exclude the override in the initial SQL Query -> Something like `DISTINCT ON (snippet_id, snippet_version)` + `ORDER BY ID` (something like that)
And then do it normally in python as you did before, without paying attention to duplicate, and if you need to disable or enable the assets, just add the overrides in the final query like
`disable this asset_id and also all other assets related to the same snippet_id+same version`

What do you think? (ignoring the fact that you would need to change the code which is annoying :grimacing: )

Reply:
I agree ! It's way better to perform these actions directly in sql, thanks !